Add support for TypeScript declarationMap

n4bb12 opened the pull request:
Adds support for declaration maps, which allows jumping to TS sources.

Resolves https://github.com/sebastian-software/preppy/issues/2.

Implementing this requires introducing a place to configure TS options. For babel, this is the project's local babel config. So it makes sense to also respect the local tsconfig for TS-related work.

The current options don't change, but they are now split into

- "defaults" (things that are optional)
- "configured" (pulled from tsconfig)
- "enforced" (things that don't make sense any other way).

Adds a declarationMap snapshot test.

n4bb12 commented:
Sidenote: prettier is not listed in devDeps, eslint doesn't match any files (not on Windows at least) so in case formatting is wrong, there is no preconfigured way to fix that.

swernerx commented on src/extractTypes.js:
```diff
@@ -34,18 +36,35 @@ function compile(fileNames, options, verbose) {
+    moduleResolution: ModuleResolutionKind.NodeJs,
+  }
+
+  const compilerOptions = merge(defaults, configured, enforced)
```
looks like a spread operation would be enough here...?

n4bb12 replied:
For declarationMap it would be enough. Some compiler options (lib, paths, rootDirs, types, typeRoots) are objects.

swernerx replied:
As we do not have them in defaults/enforced it should still work without `merge`, right?

n4bb12 replied:
right

swernerx commented on src/extractTypes.js:
```diff
@@ -34,18 +36,35 @@ function compile(fileNames, options, verbose) {
 
 export default function extractTypes({ input, root, output, verbose }) {
+  const tsConfig = join(root, "tsconfig.json")
+  const configured = existsSync(tsConfig) && require(tsConfig).compilerOptions
```
This should be probably done outside of the method to only run once.

n4bb12 replied:
It only runs once per preppy execution. So, not sure what you mean by _once_.
Also, it requires the `root` dir passed in via `options`. So, not sure where you want this to be moved to.

swernerx replied:
Mhh, in the current implementation it's required for each call of the method - which might be executed multi times in watch mode.

I would probably suggest parsing it in the same place as `package.json`, here: https://github.com/sebastian-software/preppy/blob/master/src/index.js#L32 and passing the TS configuration to the extraction method by an additional parameter.

n4bb12 replied:
https://github.com/sebastian-software/preppy/pull/3/files#diff-1fdf421c05c1140f6d71444ea2b27638R33

swernerx commented:
> Sidenotes:
> 
> Prettier is not listed in devDeps, eslint doesn't match any files (not on Windows at least). So in case formatting is wrong, there is no preconfigured way to fix that.
> 
> There are some obsolete snapshots and one test always fails due to some locking issue.

The obsolete reporting is a side-effect of the locking issue. That might just happen on AppVeyor. They still have this kind of issues from time to time. Rebuilding it typically helps: https://ci.appveyor.com/project/swernerx/preppy/builds/23449991

I cleaned up some of the linting/config issues reported. Thanks for pointing this out!

swernerx commented:
Looks like you used Prettier without the eslint postfix, right now. Also somehow the imports got sorted... is this some editor setting in your environment. They do not seem to be related to the actual changes.

swernerx commented:
Thanks BTW for your work 👍

n4bb12 commented:
Linters are still not working.
```
 npm run format

> preppy@7.6.4 format D:\Projects\github-forks\preppy
> prettier --write '**/*.js' && npm run fix

[error] No matching files. Patterns tried: '**/*.js' !**/node_modules/** !./node_modules/** !**/.{git,svn,hg}/** !./.{git,svn,hg}/**
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! preppy@7.6.4 format: `prettier --write '**/*.js' && npm run fix`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the preppy@7.6.4 format script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\ACS\AppData\Roaming\npm-cache\_logs\2019-03-29T15_56_28_412Z-debug.log
```

swernerx commented:
> Linters are still not working.
> 
> ```
>  npm run format
> 
> > preppy@7.6.4 format D:\Projects\github-forks\preppy
> > prettier --write '**/*.js' && npm run fix
> 
> [error] No matching files. Patterns tried: '**/*.js' !**/node_modules/** !./node_modules/** !**/.{git,svn,hg}/** !./.{git,svn,hg}/**
> npm ERR! code ELIFECYCLE
> npm ERR! errno 2
> npm ERR! preppy@7.6.4 format: `prettier --write '**/*.js' && npm run fix`
> npm ERR! Exit status 2
> npm ERR!
> npm ERR! Failed at the preppy@7.6.4 format script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     C:\Users\ACS\AppData\Roaming\npm-cache\_logs\2019-03-29T15_56_28_412Z-debug.log
> ```

Probably some Windows-only issue? Mhh...

n4bb12 commented:
>Also somehow the imports got sorted...

Want me to re-shuffle them?

n4bb12 commented:
This works for me:
```
npx prettier --write '**/*.js' && npx eslint --fix .
```

However it creates a

<details>
<summary>huge diff</summary>


```
babel.config.js 75ms
coverage\lcov-report\block-navigation.js 18ms
coverage\lcov-report\prettify.js 171ms
coverage\lcov-report\sorter.js 26ms
jest.setup.js 7ms
src\cli.js 8ms
src\extractTypes.js 16ms
src\getBanner.js 8ms
src\getEntries.js 16ms
src\getOutputMatrix.js 10ms
src\getRollupInputOptions.js 18ms
src\getRollupOutputOptions.js 11ms
src\getTasks.js 20ms
src\index.js 38ms
src\parseCommandline.js 15ms
src\progressPlugin.js 16ms
src\typescriptResolvePlugin.js 9ms
test\esnext-assets\dist\index.esm.js 5ms
test\esnext-assets\dist\index.umd.js 9ms
test\esnext-assets\index.test.js 6ms
test\esnext-assets\src\index.js 5ms
test\esnext-binary\bin\mycli.js 4ms
test\esnext-binary\index.test.js 6ms
test\esnext-binary\src\cli.js 5ms
test\esnext-binary\src\common.js 4ms
test\esnext-fail\index.test.js 8ms
test\esnext-fail\src\index.js
[error] test\esnext-fail\src\index.js: SyntaxError: Unexpected token, expected ";" (2:18)
[error]   1 | export default () => {
[error] > 2 |   return Math.abs)
[error]     |                  ^
[error]   3 | }
[error]   4 |
test\esnext-min\dist\index.cjs.min.js 12ms
test\esnext-min\dist\index.esm.min.js 15ms
test\esnext-min\dist\index.umd.min.js 18ms
test\esnext-min\index.test.js 12ms
test\esnext-min\src\index.js 16ms
test\esnext-multibinary\bin\first-cli.js 7ms
test\esnext-multibinary\bin\second-cli.js 6ms
test\esnext-multibinary\index.test.js 6ms
test\esnext-multibinary\src\cli\second.js 4ms
test\esnext-multibinary\src\common.js 7ms
test\esnext-multibinary\src\first.js 5ms
test\esnext\dist\index.cjs.js 15ms
test\esnext\dist\index.esm.js 11ms
test\esnext\dist\index.umd.js 15ms
test\esnext\index.test.js 9ms
test\esnext\src\index.js 11ms
test\typescript-jsx\dist\index.cjs.js 17ms
test\typescript-jsx\dist\index.esm.js 17ms
test\typescript-jsx\dist\index.umd.js 20ms
test\typescript-jsx\index.test.js 6ms
test\typescript\dist\index.cjs.js 11ms
test\typescript\dist\index.esm.js 13ms
test\typescript\dist\index.umd.js 17ms
test\typescript\index.test.js 8ms
test\universal-basic\dist\browser.esm.js 5ms
test\universal-basic\dist\browser.umd.js 5ms
test\universal-basic\dist\node.cjs.js 4ms
test\universal-basic\dist\node.esm.js 8ms
test\universal-basic\index.test.js 7ms
test\universal-basic\src\browser.js 6ms
test\universal-basic\src\index.js 4ms
test\universal-folder\bin\mycli.js 7ms
test\universal-folder\dist\browser.esm.js 4ms
test\universal-folder\dist\browser.umd.js 10ms
test\universal-folder\dist\node.cjs.js 4ms
test\universal-folder\dist\node.esm.js 4ms
test\universal-folder\index.test.js 7ms
test\universal-folder\src\cli\index.js 6ms
test\universal-folder\src\client\index.js 5ms
test\universal-folder\src\server\index.js 4ms

ACS@AZEESS MINGW64 /D/Projects/github-forks/preppy
 st
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   babel.config.js
        modified:   jest.setup.js
        modified:   src/cli.js
        modified:   src/extractTypes.js
        modified:   src/getBanner.js
        modified:   src/getEntries.js
        modified:   src/getOutputMatrix.js
        modified:   src/getRollupInputOptions.js
        modified:   src/getRollupOutputOptions.js
        modified:   src/getTasks.js
        modified:   src/index.js
        modified:   src/parseCommandline.js
        modified:   src/progressPlugin.js
        modified:   src/typescriptResolvePlugin.js
        modified:   test/esnext-assets/index.test.js
        modified:   test/esnext-assets/src/index.js
        modified:   test/esnext-binary/index.test.js
        modified:   test/esnext-binary/src/cli.js
        modified:   test/esnext-binary/src/common.js
        modified:   test/esnext-fail/index.test.js
        modified:   test/esnext-min/index.test.js
        modified:   test/esnext-min/src/index.js
        modified:   test/esnext-multibinary/index.test.js
        modified:   test/esnext-multibinary/src/cli/second.js
        modified:   test/esnext-multibinary/src/common.js
        modified:   test/esnext-multibinary/src/first.js
        modified:   test/esnext/index.test.js
        modified:   test/esnext/src/index.js
        modified:   test/typescript-jsx/index.test.js
        modified:   test/typescript/index.test.js
        modified:   test/universal-basic/index.test.js
        modified:   test/universal-basic/src/browser.js
        modified:   test/universal-basic/src/index.js
        modified:   test/universal-folder/index.test.js
        modified:   test/universal-folder/src/cli/index.js
        modified:   test/universal-folder/src/client/index.js
        modified:   test/universal-folder/src/server/index.js

no changes added to commit (use "git add" and/or "git commit -a")
```

</details>

Looks like it first needs some work on master.

swernerx commented:
> > Also somehow the imports got sorted...
> 
> Want me to re-shuffle them?

Not really... I just like to keep PRs limited to their actual scope and not touching anything outside of this.

swernerx commented:
> This works for me:
> 
> ```
> npx prettier --write '**/*.js' && npx eslint --fix .
> ```
> However it creates a
> 
> huge diff
> Looks like it first needs some work on master.

I figure I mainly used prettier only for formatting of selections inside VSCode.

n4bb12 commented:
Any more TODOs?

swernerx commented:
> Any more TODOs?

No :)
